### PR TITLE
caddyhttp: Placeholder for client cert in DER + base64 format

### PR DESCRIPTION
**Pull request by musinit**
Fixes https://github.com/caddyserver/caddy/issues/3767

### CODE REVIEW

File: modules/caddyhttp/replacer.go [hunk 1]
```diff
@@ -353,6 +354,9 @@ func getReqTLSReplacement(req *http.Request, key string) (interface{}, bool) {
 		case "client.certificate_pem":
 			block := pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}
 			return pem.EncodeToMemory(&block), true
+		case "client.certificate_pem_encoded":
+			block := pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}
+			return base64.StdEncoding.EncodeToString(pem.EncodeToMemory(&block)), true
```
Comment (mholt) [inline modules/caddyhttp/replacer.go:359]: This is essentially base64-encoding a base64-encoding of the DER bytes and wrapping it in PEM blocks. I think all we need is the base64-encoded block, without PEM-encoding.

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for the PR. I think it needs to be changed though, as it doesn't really make sense as-is.

**Comment by francislavoie** [comment]
It's not PEM if it's raw bytes base64 encoded; that's DER base64 encoded. PEM is specifically DER base64 encoded _then_ with ASCII armor added and newlines at every 64 chars.

File: modules/caddyhttp/replacer.go [hunk 2]
```diff
@@ -353,6 +354,9 @@ func getReqTLSReplacement(req *http.Request, key string) (interface{}, bool) {
 		case "client.certificate_pem":
 			block := pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}
 			return pem.EncodeToMemory(&block), true
+		case "client.certificate_pem_base64":
+			block := pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}
+			return base64.StdEncoding.EncodeToString(block.Bytes), true
```
Comment (mholt) [inline modules/caddyhttp/replacer.go:359]:
```suggestion
		case "client.certificate_der_base64":
			return base64.StdEncoding.EncodeToString(cert.Raw), true
```

**Review by mholt — CHANGES_REQUESTED** [review]
Yeah these changes aren't quite what was suggested... try something like this instead.

**Comment by mholt** [comment]
Cool, does this serve the intended purpose @musinit @mc0239 @Ramzec?

**Comment by mholt** [comment]
Someone with an actual need for this will need to verify it works and fulfills the need before this will get merged, otherwise we will have to close this for inactivity.

**Comment by mholt** [comment]
Closing due to inactivity. Can reopen if needed.

**Comment by cschmatzler** [comment]
Hey @mholt!

Not one of the people that originally wanted this, but stumbled upon it half an hour ago and can report that this change does indeed work and allows me to use Caddy in front of the Syncthing Discovery service with

```
tls {
        client_auth {
            mode require
        }
    }

	reverse_proxy /* syncthing-discovery:8443 {
		header_up X-Real-IP {http.request.remote.host}
		header_up X-Client-Port {http.request.remote.port}
		header_up X-SSL-Cert {http.request.tls.client.certificate_der_base64}
	}
```

That said, I'd appreciate you reopening and merging this PR.

Thanks!

**Comment by francislavoie** [comment]
Interesting, thanks @cschmatzler, for ref, `X-SSL-Cert` is documented here: https://docs.syncthing.net/users/stdiscosrv.html#requirements

**Comment by francislavoie** [comment]
I went ahead and rebased the branch, fixing some merge conflicts.

@mholt I think this is ready to go :+1:

**Review by mholt — APPROVED** [review]
As this has an actual (but niche) use case, I guess we can merge this. Thank you for verifying!

**Comment by aoxiangtianji** [comment]
This doesn't work on my machine.
I'm running a syncthing discovery server container and a caddy container as reverse proxy via podman rootless.
But the discovery server still reports no certificates.

This is my Caddyfile:
`{
    debug
}

sync.example.com:10443 {
    tls {
        dns cloudflare ***
        client_auth {
            mode request
        }
    }

    reverse_proxy /* syncthing-discovery:8443 {
        header_up X-Forwarded-For {http.request.remote.host}
        header_up X-Client-Port {http.request.remote.port}
        header_up X-SSL-Cert {http.request.tls.client.certificate_der_base64}
    }
}`

And the log of caddy:

`{"level":"debug","ts":1659198997.878383,"logger":"http.handlers.reverse_proxy","msg":"upstream roundtrip","upstream":"syncthing-discovery:8443","duration":0.004028418,"request":{"remote_ip":"10.89.1.15","remote_port":"55602","proto":"HTTP/1.1","method":"POST","host":"sync.aoxtj.cyou:10443","uri":"/","headers":{"X-Forwarded-Proto":["https"],"X-Forwarded-Host":["sync.aoxtj.cyou:10443"],"Content-Type":["application/json"],"Accept-Encoding":["gzip"],"X-Forwarded-For":["10.89.1.15"],"X-Client-Port":["55602"],"X-Ssl-Cert":["MIICH********************37OvaXi2w=="],"User-Agent":["Go-http-client/1.1"],"Content-Length":["133"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"","server_name":"sync.aoxtj.cyou","client_common_name":"syncthing","client_serial":"6081856432655817123"}},"headers":{"Retry-After":["1500"],"X-Content-Type-Options":["nosniff"],"Date":["Sat, 30 Jul 2022 16:36:37 GMT"],"Content-Length":["10"],"Content-Type":["text/plain; charset=utf-8"]},"status":403}`

Is there something wrong?

**Comment by francislavoie** [comment]
The headers look correct. It's not an issue with Caddy, you'll need to figure out why your upstream app doesn't accept it.

**Comment by Rirmach** [comment]
@aoxiangtianji A "tricky" solution, works for me
Just change
```
header_up X-SSL-Cert {http.request.tls.client.certificate_der_base64}
```
to
```
header_up X-SSL-Cert "-----BEGIN CERTIFICATE-----%0a{http.request.tls.client.certificate_der_base64}%0a-----END CERTIFICATE-----"
```
Where %0a stands for the url-encoded line break

**Pull request merged**